### PR TITLE
Add aten mkldnn batchnorm backward operator

### PR DESCRIPTION
**Pull request by XiaobingSuper**
### mkldnn backward ops list:
 - [ ] \(https://github.com/pytorch/pytorch/pull/20567) Add aten mkldnn conv2d backward operator :yellow_heart:
 - [ ] \(https://github.com/pytorch/pytorch/pull/20570) Add aten mkldnn backward ops: relu, linear and reshape :yellow_heart:
 - [ ] \(https://github.com/pytorch/pytorch/pull/20571) Add aten mkldnn backward ops: max_pool2d, avg_pool2d and adaptive_avg_poo2d :yellow_heart:
 - [ ] \(https://github.com/pytorch/pytorch/pull/20572) Add aten mkldnn batchnorm backward operator :yellow_heart:
 - [x] \(https://github.com/pytorch/pytorch/pull/20573) Add aten mkldnn zero_ operator💚 
- [x] \(https://github.com/pytorch/pytorch/pull/20575) Add mkldnn mul operator 💚

### CODE REVIEW

**Comment by VitalyFedyunin** [comment]
Can you please rebase on top of the master

**Comment by XiaobingSuper** [comment]
@VitalyFedyunin

File: test/test_mkldnn.py [hunk 1]
```diff
@@ -182,16 +182,45 @@ def test_batch_norm2d(self):
...
+                        mkldnn_bn = mkldnn_utils.to_mkldnn(copy.deepcopy(bn))
+                    self.assertEqual(
+                        bn(x),
+                        mkldnn_bn(x.to_mkldnn()).to_dense())
+                    if (not train and track_running_stats):
+                        self._test_serialization(mkldnn_bn, (x.to_mkldnn(),))
+                        self._test_tracing(mkldnn_bn, (x.to_mkldnn(),))
+
+    def test_batch_norm2d_backward(self):
+        N = torch.randint(3, 10, (1,)).item()
```
Comment (VitalyFedyunin) [inline test/test_mkldnn.py]: Please do not relay on random numbers. You are not freezing random seed, that means every test execution will generate new shape. It might create false sense of security, but many modern test frameworks will attempt to rerun failed test hiding problem prom developers.

Much better if you explicitly specify shapes you are interested to test.

File: aten/src/ATen/native/mkldnn/Normalization.cpp [hunk 2]
```diff
@@ -39,38 +53,86 @@ std::tuple<Tensor, Tensor, Tensor> mkldnn_batch_norm(
     bool train,
     double momentum,
     double eps) {
+  AT_ASSERTM(input.dim() == 4 || input.dim() == 5,
```
Comment (VitalyFedyunin) [inline aten/src/ATen/native/mkldnn/Normalization.cpp]: `AT_ASSERTM` is deprecated, please use `TORCH_INTERNAL_ASSERT`

File: aten/src/ATen/native/mkldnn/MKLDNNCommon.cpp [hunk 3]
```diff
@@ -75,6 +75,19 @@ ideep::tensor itensor_view_from_dense(const Tensor& tensor) {
...
 }
+
+// Note in case the aten Tensor is a dense tensor, the retured ideep
+// tensor is just a view of the storage of the aten dense tensor, so
+// caller needs to make sure the aten dense tensor's lifetime is
+// longer than the ideep tensor.
+ideep::tensor get_mkldnn_tensor(const at::Tensor& tensor) {
+  if (tensor.is_mkldnn()) {
+    return at::native::itensor_from_mkldnn(tensor);
+  } else {
```
Comment (VitalyFedyunin) [inline aten/src/ATen/native/mkldnn/MKLDNNCommon.cpp]: What if tensor is sparse? We need to error out.

File: aten/src/ATen/native/mkldnn/MKLDNNCommon.h [hunk 4]
```diff
@@ -30,6 +30,10 @@ ideep::tensor& itensor_from_mkldnn(const Tensor& mkldnn_tensor);
 // Construct an `ideep::tensor` "view" from dense tensor, note the
 // ideep::tensor will share the underlying buffer
 ideep::tensor itensor_view_from_dense(const Tensor& tensor);
+
+// Helper function for getting an ideep tensor out of an aten Tensor.
+ideep::tensor get_mkldnn_tensor(const at::Tensor& tensor);
```
Comment (VitalyFedyunin) [inline aten/src/ATen/native/mkldnn/MKLDNNCommon.h]: Reading below, this function name is terribly confusing, as you are changing `itensor_from_mkldnn` to `get_mkldnn_tensor` which now reads as, I want mkldnn_tensor as result. I suggest to rename it to `itensor_from_tensor`.

File: aten/src/ATen/native/mkldnn/Normalization.cpp [hunk 5]
```diff
@@ -20,6 +20,20 @@ std::tuple<Tensor, Tensor, Tensor> mkldnn_batch_norm(
...
+    const Tensor& input,
+    const Tensor& weight,
+    const Tensor& running_mean,
+    const Tensor& running_var,
+    const Tensor& save_mean,
+    const Tensor& save_invstd,
+    bool train,
+    double eps,
+    std::array<bool,3> grad_input_mask) {
+  AT_ERROR("mkldnn_batch_norm_backward: ATen not compiled with MKLDNN support");
```
Comment (VitalyFedyunin) [inline aten/src/ATen/native/mkldnn/Normalization.cpp]: // Deprecated alias; this alias was deprecated because it represents extra API
// surface that makes it hard for people to understand what macro to use.
// Use TORCH_CHECK(false, ...) or TORCH_INTERNAL_ASSERT(false, ...) to
// unconditionally fail at a line of code.

File: aten/src/ATen/native/mkldnn/Normalization.cpp [hunk 6]
```diff
@@ -39,38 +53,86 @@ std::tuple<Tensor, Tensor, Tensor> mkldnn_batch_norm(
...
+    const Tensor& input,
+    const Tensor& weight,
+    const Tensor& running_mean,
+    const Tensor& running_var,
+    const Tensor& save_mean,
+    const Tensor& save_invstd,
+    bool train,
+    double eps,
+    std::array<bool,3> grad_input_mask) {
+  AT_ASSERTM(train, "mkldnn_batch_norm_backward: currently mkldnn only support train model");
```
Comment (VitalyFedyunin) [inline aten/src/ATen/native/mkldnn/Normalization.cpp]: `TORCH_INTERNAL_ASSERT`

File: aten/src/ATen/native/mkldnn/Normalization.cpp [hunk 7]
```diff
@@ -39,38 +53,86 @@ std::tuple<Tensor, Tensor, Tensor> mkldnn_batch_norm(
...
+  ideep::tensor& x = itensor_from_mkldnn(input);
+  ideep::tensor w = itensor_view_from_dense(weight);
+  ideep::tensor& m = itensor_from_mkldnn(save_mean);
+  ideep::tensor& v = itensor_from_mkldnn(save_invstd);
+
+  ideep::tensor gradx, gradw, gradb;
+  ideep::batch_normalization_backward::compute<AllocForMKLDNN>(
+      x, m, v, grady, w, gradx, gradw, gradb, eps);
+
+  if (weight.is_mkldnn()) {
```
Comment (VitalyFedyunin) [inline aten/src/ATen/native/mkldnn/Normalization.cpp]: Ok if weigh could be is_mkldnn, how line 115 supposed to work `ideep::tensor w = itensor_view_from_dense(weight);`

**Review by VitalyFedyunin — CHANGES_REQUESTED** [review]

File: aten/src/ATen/native/mkldnn/MKLDNNCommon.cpp (hunk 3 above)
Comment (XiaobingSuper) [reply aten/src/ATen/native/mkldnn/MKLDNNCommon.cpp]: **itensor_view_from_dense** will check whether the tensor is dense, if not, will error out.

**Comment by XiaobingSuper** [comment]
@VitalyFedyunin,  I change the code according to your suggestions, thanks!

**Comment by XiaobingSuper** [comment]
@VitalyFedyunin , the two failed cases is about cuda build,  which are not caused by this changes, please help see tnem, thanks!

**Comment by XiaobingSuper** [comment]
@VitalyFedyunin

**Comment by XiaobingSuper** [comment]
@VitalyFedyunin

**Comment by XiaobingSuper** [comment]
@VitalyFedyunin

File: aten/src/ATen/native/mkldnn/Normalization.cpp [hunk 8]
```diff
@@ -39,38 +53,87 @@ std::tuple<Tensor, Tensor, Tensor> mkldnn_batch_norm(
     bool train,
     double momentum,
     double eps) {
+  TORCH_CHECK(input.dim() == 4 || input.dim() == 5,
```
Comment (VitalyFedyunin) [inline aten/src/ATen/native/mkldnn/Normalization.cpp]: No test case for 3d

**Review by VitalyFedyunin — APPROVED** [review]

**Comment by VitalyFedyunin** [comment]
Do you have specific subset of operators in mind to add mkldnn backwards support?

File: aten/src/ATen/native/mkldnn/Normalization.cpp (hunk 8 above)
Comment (XiaobingSuper) [reply aten/src/ATen/native/mkldnn/Normalization.cpp]: 3d operators will be enable at next step, including conv3d, pool3d, and other mkldnn operators.

**Comment by XiaobingSuper** [comment]
Now, we want to enable conv, pool, batchnorm. linear, relu and reshape operators, which can run resnet and resnext models, and then add other operators for user's requests, such as 3d operators for Unet , laynorm, concat operators for transformer models. Thanks!

**Comment by XiaobingSuper** [comment]
@VitalyFedyunin, Can you help to merge this PR? Thanks!
